napari/_qt/threading.py

```diff
@@ -0,0 +1,563 @@
+
+    def quit(self) -> None:
+        """Send a message to abort the worker."""
+        self._abort_requested = True
+
+    @property
+    def abort_requested(self) -> bool:
+        """Whether the worker has been requested to stop."""
+        return self._abort_requested
+
+    @property
+    def is_running(self) -> bool:
+        """Whether the worker has been started"""
+        return self._running
+
+    @Slot()
+    def run(self):
+        """Start the worker.
+
+        **This** is the function that actually gets called when calling
+        :func:`start_worker(worker)`.  It simply wraps the :meth:`work` method,
+        and emits a few signals.  Subclasses should NOT override this method
+        (except with good reason), and instead should implement :meth:`work`.
+        """
+        self.started.emit()
+        self._running = True
+        try:
+            self.work()
+        except Exception as exc:
+            self.errored.emit(exc)
+        self.finished.emit()
+
+    def work(self):
+        """Main method to execute the worker.
+
+        Subclasses must implement this method.  Minimally, it should check
+        ``self.abort_requested`` periodically and exit if True.
+
+        Example
+        -------
+
+        .. code-block:: python
+
+            def work(self):
+                i = 0
+                while True:
+                    if self.abort_requested:
+                        self.aborted.emit()
+                        break
+                    i += 1
+                    if i > max_iters:
+                        break
+                    time.sleep(0.5)
+
+        """
+        raise NotImplementedError(
+            f'"{self.__class__.__name__}" failed to define work() method'
+        )
+
+
+class Worker(WorkerBase):
+    """QRunnable with signals that wraps a long-running function.
+
+    When combined with :func:`new_worker_qthread`, provides a convenient way
+    to run a function in another thread, while allowing 2-way communication
+    between threads, using plain-python generator syntax in the original
+    function.
+
+    Parameters
+    ----------
+    func : callable
+        The function being run in another thread.  May be a generator function.
+    *args
+        Will be passed to func on instantiation
+    **kwargs
+        Will be passed to func on instantiation
+    """
+
+    def __init__(self, func: Callable, *args, **kwargs):
+        super().__init__()
+        if inspect.isgeneratorfunction(func):
+            self._gen = func(*args, **kwargs)
+        else:
+            self._gen = as_generatorfunction(func)(*args, **kwargs)
+
+        self._incoming_value = None
+        self._pause_requested = False
+        self._resume_requested = False
+        self._paused = False
+        self._pause_interval = 0.01
+        self._running = False
+
+        first_yield = next(self._gen)
+        if isinstance(first_yield, dict):
+            self.parse_first_yield(first_yield)
```

Comment:
Set "setup in the main thread before calling yield" is definitely confusing to me. Just having a function that runs in two different threads is already confusing. And having that transition be controlled by "the first yield" seems magic and unconventional? Or is this in fact a pattern from somewhere?

It's weird that line 1 of the function will start in a different thread based on something that comes later in the function (whether it's a generator). Once you get it it makes sense, but it is kind of hidden. I guess the way I'd illustrate that it's hidden is I imagine every function every user writes will probably contain a giant comment "this part runs in the main thread" and "now we are in the worker thread" to explain it. The need for those comments to me is an illustration things are kind of hidden.

I think it's very compact and clever. It's a clever way to fit in a lot of functionality using implicit built-in stuff, without anything explicit. But that's the drawback as well. I think we are very proud to say "this is really simple, you just define one function", but in fact it's not actually simple.

I really don't know a better way though. Not off hand I'd have to think about it. My guess is you'd get rid of the idea of having the user just define a function. And you'd have a base class people inherit from. Then you can make it clear which method runs in which thread, etc. Make everything way more explicit. But less compact/clever. Or maybe there is a model we can find of doing all this in a standard way?

This would work and it's totally learnable. I think people would figure it out from the documentation. But it always seem a bit magic/custom I think.

Reply:
My comment above is a little out of date now: After @AhmetCanSolak's review, I made that setup strictly opt-in (with the `init_yield=True` option).  I agree it's too implicit to be the default.  I can also take it out all together as well and I wouldn't miss it too much.  Is this whole comment just about that first yield ?  or are you talking more broadly?

Reply:
I guess `init_yield=True` helps. In that case though, when True, I just feel it's very hidden that the one function will be called from two different threads. Unless it's some sort of convention or pattern people do with generators?

Basically we're saying in the user's function yield doesn't just mean "yield" it means "yield and, if this is the first yield, call me in another thread from now on". To me that's just inherently hidden/implicit. In the user's function there's no indication that anything with threading happening, yet it is totally happening, and in fact it can be critical to know which thread is running, but that's basically only evident in comments.

I'm not sure what else I'd suggest. I feel like it was strong goal here that user provide a **function** and not have `WorkerThread` base class? Given that constraint, this is a clever solution I guess. If it were just internal code, not providing an API for users, I would never make that constraint, that it has to be just a function. I'd probably have a base class with a super explicit `setup()` method with a big comment "This runs in the main thread" and a `run()` method that "runs in the worker thread".

But that's way clunkier. So maybe the trade off here is reasonable. There's a bit of magic but it's compact. I'm not sure, I would just never have gone down this road myself, but I see the benefits.

Reply:
> I feel like it was strong goal here that user provide a function and not have WorkerThread base class

Yeah, I think that is a goal here.  I don't know whether this is an "explicit" goal of napari, but I think we generally try to offer at least one API that relative python newcomers (e.g. scientists who don't have the time to learn about class inheritance) could take advantage of to accomplish the majority of their needs in a relatively efficient/safe way.  This certainly doesn't prevent people from subclassing `WorkerBase` and taking it in any direction... and some may even use the `init_yield` in their subclasses (as I did with the `ProgressWorker`).  And since the default is `init_yield==False`, it's as if it wasn't there for users who don't want it.

But your point is well taken!  There's nothing that couldn't be done here in a more explicit way in the main thread by the end-user.  I'm not too married to this pattern (i kinda like it, but I don't _love_ it), and happy to keep or remove it according to the group consensus.